contracts/staking/StakingBase.sol

```diff
@@ -591,21 +642,211 @@ abstract contract StakingBase is ERC721TokenReceiver {
+            amounts = new uint256[](totalNumReceivers);
+
+            // Default setup implies that all bonds are equal
+            // Get each operator reward
+            uint256 operatorReward = reward / (totalNumReceivers);
+
+            // Get corresponding operators and set operators reward amounts
+            for (uint256 i = 0; i < numInstances; ++i) {
+                receivers[i] = IService(serviceRegistry).mapAgentInstanceOperators(agentInstances[i]);
+                amounts[i] = operatorReward;
+            }
+
+            // Set service owner address and its reward amount
+            receivers[numInstances] = serviceOwner;
+            // Service owner gets its reward amount and a division remainder, if any
+            amounts[numInstances] = reward - (numInstances * operatorReward);
+        } else if (rewardDistributionType == RewardDistributionType.ServiceOwner) {
+            // Allocate arrays
+            receivers = new address[](1);
+            amounts = new uint256[](1);
+
+            // Set service owner address and its reward amount
+            receivers[0] = serviceOwner;
+            amounts[0] = reward;
+        } else if (rewardDistributionType == RewardDistributionType.ServiceMultisig) {
+            // Allocate arrays
+            receivers = new address[](1);
+            amounts = new uint256[](1);
+
+            // Set service owner address and its reward amount
+            receivers[0] = multisig;
+            amounts[0] = reward;
+        } else {
+            // Custom rewards distributor address: shift rewardDistributionType value of 8 bits
+            // Note this address was already checked for not being zero
+            address customRewardsDistributor = address(uint160(rewardDistributionInfo >> 8));
+
+            // Get receivers and amounts from external customRewardsDistributor contract
+            (receivers, amounts) = ICustomRewardsDistributor(customRewardsDistributor).getRewardReceiversAndAmounts(serviceId,
+                serviceOwner, multisig, reward);
+
+            // Sum all calculated amounts
+            uint256 amountCheck;
+            for (uint256 i = 0; i < amounts.length; ++i) {
+                amountCheck += amounts[i];
+            }
+
+            // Check for reward amount to match total calculated amount
+            if (amountCheck != reward) {
+                revert WrongAmount(amountCheck, reward);
+            }
+        }
+    }
+
+    /// @dev Stakes the service.
+    /// @notice Each service must be staked for a minimum of maxInactivityDuration time, or until the funds are not zero.
+    ///         maxInactivityDuration = maxNumInactivityPeriods * livenessPeriod
+    /// @param serviceId Service Id.
+    /// @param rewardDistributionInfo Reward distribution info: rewardDistributionType and customRewardsDistributor
+    ///        address, if required.
+    function _stake(uint256 serviceId, uint256 rewardDistributionInfo) internal {
```

Comment:
This was just the code shuffling around, the majority of this function is left untouched. Will mention below what was added.